docs/proposals/20191016-e2e-test-framework.md

```diff
@@ -0,0 +1,235 @@
+
+This framework will be implemented using the Ginkgo/Gomega behavioral testing framework because that is the Kubernetes
+standard and does a good job of structuring and organizing behavior based tests.
+
+This framework will live in the `sigs.k8s.io/cluster-api` module because it will be formalizing behaviors of Cluster
+API. Therefore the set of behaviors for each version may change, so will the test suite.
+
+This framework is not conformance nor will be built in the same style as the conformance framework of kubernetes. That
+is to say, there will not be a distributed binary. The only artifact needed is Cluster API's go module appropriately
+tagged and pushed.
+
+These tests will be consumed by a provider just like any other go library dependency. They can be imported and provider
+custom types can be passed in and the e2es will be run for the given provider, assuming secrets are also managed
+
+The exact interface that a provider must satisfy is to be determined by what the tests demand. The plan for
+implementation at a high level is to build out reusable but pluggable testing components. For example, the framework
+will need a way to start a management cluster. We will use KIND initially and abstract the actual details away and
+expose an interface. If you want to use a different style of management cluster, say, a cloud-formation based cluster,
+then you will have to wrap up the cloud-formation code in some object and implement the testing framework’s exposed
+interface for the management cluster and use your object where the sample code uses the KIND-based management cluster
+struct. I expect but cannot guarantee that there will be interfaces for each provider but exact details are to be
+determined as the implementation gets more fleshed out.
+
+#### Behaviors to test
+
+* There is a Kubernetes cluster with one node that passes a healthcheck after creating a properly configured Cluster,
+  InfraCluster, Machine, InfraMachine and BootstrapConfig.
+* Creating the resources necessary for three control planes will create a three node cluster.
+* Deleting a cluster deletes all resources associated with that cluster including Machines, BootstrapConfigs,
+  InfraMachines, InfraCluster, and generated secrets.
+* Creating a cluster with one control plane and one worker node will result in a cluster with two nodes.
+* The version fields in Machines are respected within the bounds of the Kubernetes skew policy.
+* Creating a control plane machine and a MachineDeployment with two replicas will create a three node cluster with one
+  control plane node and two worker nodes.
+* Modifying a replica count on a MachineDeployment will modify the number of worker nodes and Machines in running state
+  the cluster has.
+* Optionally, Machines report failures when their underlying InfraMachines report failures.
+* Manage multiple workload clusters in different namespaces.
+* Workload Clusters created pass Kubernetes Conformance.
+
```

Comment:
Would testing that a given Machine part of a MachineDeployment/MachineSet will get recreated if it's manually deleted be something that would be included here or is that more CAPI specific?